vscode/src/format.ts

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { ILanguageService } from "qsharp-lang";
+import * as vscode from "vscode";
+import { toVscodeRange } from "./common";
+import { EventType, sendTelemetryEvent } from "./telemetry";
+import { getRandomGuid } from "./utils";
+
+export function createFormatProvider(languageService: ILanguageService) {
+  return new QSharpFormatProvider(languageService);
+}
+
+class QSharpFormatProvider implements vscode.DocumentFormattingEditProvider {
+  constructor(public languageService: ILanguageService) {}
+
+  async provideDocumentFormattingEdits(document: vscode.TextDocument) {
+    // telemetry start format
+    const associationId = getRandomGuid();
+    sendTelemetryEvent(EventType.FormatStart, { associationId }, {});
```

Comment:
I love that you thought of telemetry, but what exactly is the goal of formatting telemetry? I can't imagine formatting possibly being a performance burden worthy of tracking --like, I almost think sending the telemetry event itself might take longer than traversing a document and collecting format edits.

Reply:
We want to be able to track how often the feature is used to learn how effective it is.